### PR TITLE
Fix security vulnerability

### PR DESCRIPTION
**Pull request by litzinger**

### CODE REVIEW

File: system/expressionengine/third_party/mo_variables/ext.mo_variables.php [hunk 1]
```diff
@@ -322,6 +318,23 @@ protected function set_global_var($key, $value = '', $xss_clean = FALSE, $embed
...
+     */
+	private function clean($value)
+    {
+    	if (APP_VER > 3) {
+			$value = ($xss_clean) ? ee('Security/XSS')->clean($value) : $value;
+		} else {
+			$value = ($xss_clean) ? $this->EE->security->xss_clean($value) : $value;
+		}
+
+        $value = str_replace(array('{', '}'), '', $value);
```
Comment (rsanchez) [inline system/expressionengine/third_party/mo_variables/ext.mo_variables.php]: Can you replace with HTML entities instead?

```
$value = str_replace(array('{', '}'), array('&#123;', '&#125;'), $value);
```

**Comment by rsanchez** [comment]
This looks good other than the one comment I left.

Let me know if you have interest in becoming a co-maintainer. I'm not doing much EE3 work now.

**Comment by croxton** [comment]
Hey Rob, encoding those braces doesn't work, the tag still parses!

**Comment by rsanchez** [comment]
That's awful. @croxton Do you know if EE specifically parses the literal string `&#123;` or if it somehow HTML decodes it to `{` and then parses it?

If it's not possible to print curly braces without EE trying to parse them, then I suppose the original implementation here (removing them altogether) is the best solution.

**Comment by croxton** [comment]
Brian, that clean function needs to be:

```
   /**
     * @param $value
     * @return mixed
     */
   private function clean($value)
    {
    	if (APP_VER > 3) {
			$value = ee('Security/XSS')->clean($value);
		} else {
			$value = $this->EE->security->xss_clean($value);
		}

        $value = str_replace(array('{', '}'), '', $value);

        return $value;
    }
```

**Comment by croxton** [comment]
Rob, i'm not sure. I think it is necessary to strip `{` and `}`. At this point I don't think we need to strip the encoded versions of these characters, still testing.

**Comment by croxton** [comment]
Passing `%7B` in the uri in place of `{` and `%7D` in place of `}` also leads to them being parsed. But the strip of `{` and `}` does stop that. So Brians approach to strip `{` and `}` is necessary.

**Comment by croxton** [comment]
I can confirm EE templates only parses the literal `{` and `}` characters, the html or url encoded forms do not work. So somewhere the encoded form passed in the url must be decoded.

**Comment by croxton** [comment]
Sorry, ignore me. I mistyped one of the encoded characters in my test.

Converting to encoded characters DOES work to make the input safe.

**Comment by litzinger** [comment]
Updated the PR. Spacing is wonky even though it looks correct locally ¯\_(ツ)_/¯

**Comment by litzinger** [comment]
Rob, you can make me a co-maintainer. I can monitor PRs and stuff.

**Comment by croxton** [comment]
Ok, so that deals with EE tags. But XSS still works > #47

**Comment by litzinger** [comment]
Right, which is odd b/c its going through xss_clean. So is that a failure in EE's core code?

**Comment by croxton** [comment]
Seems so.

**Comment by croxton** [comment]
FWIW, this stops the XSS as well:

```
    /**
     * @param $value
     * @return mixed
     */
    private function clean($value)
    {
    	if (APP_VER > 3) {
	    $value = ee('Security/XSS')->clean($value);
	} else {
	    $value = $this->EE->security->xss_clean($value);
	}

        $value = htmlspecialchars($value);
        $value = str_replace(array('{', '}'), array('&#123;', '&#125;'), $value);
        
        return $value;
    }
```

**Comment by rsanchez** [comment]
This should do it:

```
$value = htmlentities($value, ENT_QUOTES, ee()->config->item('charset') ?: 'UTF-8');
```

**Comment by croxton** [comment]
OK seems that `Security/XSS` removes scripts, images and a lot of other XSS attacks, but is not actually encoding all HTML, only what it has to (`<script>` and the like). Which is right I guess. It just so happens that in this instance it is necessary to filter the user input further.

**Comment by croxton** [comment]
Yep @rsanchez I can confirm that is working great.


	/**
	 * @param $value
	 * @return mixed
	 */
	private function clean($value)
	{
		if (APP_VER > 3) {
			$value = ee('Security/XSS')->clean($value);
		} else {
			$value = $this->EE->security->xss_clean($value);
		}

		$value = htmlentities($value, ENT_QUOTES, ee()->config->item('charset') ?: 'UTF-8');
		$value = str_replace(array('{', '}'), array('&#123;', '&#125;'), $value);
		
		return $value;
	}

**Comment by croxton** [comment]
Good to go I think.

**Pull request merged**